Introduce bin/setup and basic structure of README.md,

zekefast opened the pull request:
# Overview

Introduce `bin/setup` script which is able to create symbolic links from configuration in ` CONFIG/home/USER_PLACEHOLDER/.config/` of thie repository to `~/.config/` on target system.

# Changes

- Introduce `bin/setup`
- Add new sections to `README.md`
- Introduce empty directory for configuration files `CONFIG`.

# Usage

```bash
$ cd <project_directory>

# Watch script documentation
$ bin/setup -h

# Run creation of symbolic links (which do nothing at the moment)
$ bin/setup
```

zekefast committed:
Add blank slate for setup script - bin/setup.

Changes:
- bin/setup at the moment contains just useful for Bash development goodies:
  - head comments
  - Bash settings through "set -o" command
  - paths to directories (__DIR__, __LIB__)
  - main and usage functions with basic help information and support
    for -h option.

zekefast committed:
Add bin/setup description to Scripts section in README.md.

zekefast committed:
Add ability to configure owner of simlinked configuration.

Changes:
- Introduce "DEFAULT_USER" constant which is determined based on $USER
  variable received from execution environment.
- Add support for -U option with obligatory parameter to redefine
  owner of symlinked configuration files and directories.
- Describe -U option in script's usage documentation (available through -h
  option).

zekefast committed:
Add ability to set home directory to which symlink configuration to.

Changes:
- Introduce DEFAULT_HOME_DIR constant which is determined based on $HOME
  variable received from execution environment.
- Add support for -H option with obligatory parameter to redefine
  home directory of symlinked configuration.
- Describe -H option in script's usage documentation (available through -h
  option).

zekefast committed:
Declare __CONFIG__ and USER_HOME_DIR_PLACEHOLDER const in bin/setup.

__CONFIG__ holds path to configuration files which are expected to be
symlinked to target system.

USER_HOME_DIR_PLACEHOLDER is a name of user's home directory under
CONFIG/.

Additional changes:
- Introduce empty directory CONFIG/home/USER_PLACEHOLDER/.config/ to hold
  user's configuration.

zekefast committed:
Make USER_NAME and HOME_DIR_PATH in bin/setup global and mutable.

As default values of USER_NAME and HOME_DIR_PATH could be changed through
-U and -H options in main function I made them mutable and globally
accessable to avoid passing them as parameters to functions which may
use them.

Changes:
- Rename constant: DEFAULT_HOME_DIR -> DEFAULT_HOME_DIR_PATH.
- Rename local variable in main function _USER to USER_NAME and make it
  global instead.
- Rename local variable in main function _HOME_DIR to HOME_DIR_PATH and
  make it global instead.
- Remove variable initialization code in main function.
- Introduce empty do_symlinking function and call it from main.

zekefast committed:
Introduce symlink_configuration function in bin/setup.

symlink_configuration function symlinking configuration from repository
to ~/.config directory on target system.

It is idempotent, i.e. save to call several times! In case of presence
of target function outputs messages "No symlinking occured! Target
<TARGET_PATH_HERE> exists!".

zekefast committed:
Add new, mostly empty sections to README.md.

Changes:
- Add new sections:
  - Supported Target Architectures
  - Supported Target OSes
  - List of Installed and Configured Software
    - Installaled
    - Configured

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+#
+# Script is idempotent. It is same to run it several times.
+#
+# TODO:
```
TODO what?

zekefast replied:
I'll fill it later a bit. It is to track scripts things which accumulated during development.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+
+
+readonly __DIR__="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+readonly __LIB__="${__DIR__}/../lib-bash"
```
Looks like `lib-bash` does not exist at the moment

zekefast replied:
I had to add `.keep` file into it! Good catch!

zekefast replied:
Added by 3bbd769999cc62793f9e4f4d7501a4c389967809

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+#set -o xtrace
+
+
+readonly __DIR__="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Looks like `readonly` is not supported in Bash 3, that means that your script will not work on OS X :+1:

zekefast replied:
I do not plan to support it! But thanks for notice! I didn't know that it is not supported on OS X.

zekefast replied:
I added dependencies version declaration in d304cff5cff22956334b8b271afa8468821156c2.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+
+readonly USER_HOME_DIR_PLACEHOLDER="USER_PLACEHOLDER"
+
+readonly DEFAULT_USER="$USER"
```
Double quotes are not necessary

zekefast replied:
Addressed by 4532f2df22a5565f858d084737526a814d8c419f.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+readonly USER_HOME_DIR_PLACEHOLDER="USER_PLACEHOLDER"
+
+readonly DEFAULT_USER="$USER"
+declare  USER_NAME="${DEFAULT_USER}"
```
`declare` is not supported in Bash 3: https://www.tldp.org/LDP/abs/html/bashver3.html

zekefast replied:
I think that's fine. As I mostly target Linux which has 4.x bash support. When I decide to port it to OS X if at all I'll change and test that later.

zekefast replied:
I added dependencies version declaration in d304cff.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+readonly DEFAULT_USER="$USER"
+declare  USER_NAME="${DEFAULT_USER}"
+
+readonly DEFAULT_HOME_DIR_PATH="$HOME"
```
Double quotes are not necessary

zekefast replied:
Addressed by 4532f2df22a5565f858d084737526a814d8c419f.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+declare  HOME_DIR_PATH="${DEFAULT_HOME_DIR_PATH}"
+
+readonly DEFAULT_CONFIG_DIR_NAME=".config"
+declare  CONFIG_DIR_NAME="${DEFAULT_CONFIG_DIR_NAME}"
```
Same thing with double quotes

zekefast replied:
Addressed by 4532f2df22a5565f858d084737526a814d8c419f.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+        USER_NAME="$OPTARG"
+        ;;
+      H)
+        HOME_DIR_PATH="$OPTARG"
```
Same regarding double quotes:) (and 3 lines above)

zekefast replied:
Addressed by 4532f2df22a5565f858d084737526a814d8c419f.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+
+  do_symlinking
+
+  exit 0
```
I think exit status will `0` by default.
I assume your intention is to make `exit 0` very explicit.

zekefast replied:
I probably will go with default as you commented that. :)

zekefast replied:
Addressed by 71fbd1897e8474e00f1a49a900ae21711931b0aa.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+  cat <<-EOS
+
+  Setup script to symlink configuration and dot files from this repository
+  to proper places in the system.
```
This partially repeats what is written in the comment.
If you want to look very cool :sunglasses: , you can build complete help message from the documentation.
See example here: https://github.com/greyblake/beep-alarm/blob/master/bin/beep-alarm#L175

zekefast replied:
It is very cool trick! But it seems a bit hacky! I leave my approach as yours wouldn't work for me because I need precise control over how my documentation looks like and I want to interpolate variables in it which means it can not go into comments at the header of the script.

greyblake commented on bin/setup:
```diff
@@ -0,0 +1,127 @@
+  local APP_NAME="$1"
+  local APP_CONFIG_DIR_NAME="${2:-$APP_NAME}"
+
+
```
2 new lines?

zekefast replied:
Yep. This my style to separate parameters in Bash which makes code more readable.

zekefast committed:
Add empty lib-bash directory.

zekefast committed:
Remove unnecessary double quotting on variable assignments.

zekefast committed:
Fallback to default exit status 0 instead of setting it explicitly.

zekefast committed:
Introduce DEPENDENCIES.md to declare software and hardware requirements.

Changes:
- Add DEPENDENCIES.md to declare software and hardware requirements of
  scripts and software in this repository.
- Move architecture and OS requirements from README.md to
  DEPENDENCIES.md.

zekefast commented:
@greyblake I addressed most of your comments! Could you check it once again? Thank you in advance!

greyblake commented:
:+
